uji.py

```diff
@@ -1054,6 +1075,159 @@ def upload(self):
         except Exception as e:
             logger.error(f'Failed to copy {filename}: {e}')
 
+    def remove_code_block_content(self, lines, from_offset=0):
+        new_lines = lines[:from_offset + 1]
+        codeblock_offset = -1
+        in_codeblock = False
+        for line in lines[from_offset + 1:]:
+            if in_codeblock:
+                if '```' in line:
+                    in_codeblock = False
+                else:
+                    # remove existing output
+                    continue
+            if codeblock_offset < 0:
+                # find the existing ``` markers
+                if '```' in line:
+                    in_codeblock = True
+                    codeblock_offset = len(new_lines) + 1
+                # new codeblock, enforce our markers, right below cursor_offset
+                if self.is_checkbox(line):
+                    codeblock_offset = self._insert(self.cursor_offset + 1, '```\n', target=new_lines)
+                    self._insert(codeblock_offset, '```\n', target=new_lines)
+            # rest is going to be forwarded as it was
+            new_lines.append(line)
+
+        # return the index where to store the output of the command
+        return new_lines, codeblock_offset
+
+    def _sanitize(self, line):
+        return line.replace('\t', '    ')
+
+    def execute_command(self):
+        line = self.lines[self.cursor_offset]
+        if not self.is_checkbox(line) or "⚙" not in line:
+            return
+
+        # list of regex to deduce the type of command to be run
+        command_re = {
+                'attach': r'^(\s*)- \[.\].* \[`(.*)`\]\((.*)\).*',
+                'human': r'^(\s*)- \[.\].* `(.*)`: <strong>(.*)</strong>$',
+                'single': r'^(\s*)- \[.\].* `(.*)`: `(.*)`$',
+                'multi': r'^(\s*)- \[.\].* `(.*)`:$',
+                'exitcode': r'^(\s*)- \[.\].* `(.*)`$',
+        }
+
+        command_type = None
+        command_match = None
+
+        for type_c, re_c in command_re.items():
+            match = re.match(re_c, line)
+            if match:
+                command_type = type_c
+                command_match = match
+                break
+
+        if command_type is None:
+            logger.error(f'Failed to match run command line: {line}')
+            return
+
+        command = match[2]
+        output = []
+        insert_offset = self.cursor_offset + 1
+        offset = -1
+
+        if command_type == 'multi':
+            self.lines, offset = self.remove_code_block_content(self.lines, self.cursor_offset)
+            self.writeout()
+            self._redraw()
+            self._render()
+
+
```

Comment:
```suggestion

```
(remove duplicate empty line)

Reply:
> Having said that, to test the above I wrote a minimal test script and it didn't behave at all what I expected. First, `proc.stdout.readlines()` blocks, so this loop can't be entered more than once anyway.

Well, the loop is only for the case `multi`.

Basically, what I want (and coded) is:
- if the run is `single`, `human`, or `exitcode`, we just run the process in the background, wait for it to terminate and then get the stdout/stderr.
  - if `single` is in use, we check if there are one or more outputs in stdout, and either append the output to the run command, either open a new code block with *all* the output
- if the run is `multi`, then we pipe the output to the screen, and we refresh the screen whenever we get a new line

So in the first case, using `readlines()` is fine, because we don't care about getting lines one by one.

And in the second case, we use `readline()`, which raises the python script at every flush of a new line. (which is what was missing in `libinput debug-events`: it doesn't flush stdout.

> Second, the [docs](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.stderr) have a warning that we can deadlock if the pipe is full. Unlikely for our use-case, but it suggests to use `popen.communicate` instead.

Yeah. If you write the test to be a `single` command and deadlock because you get too many outputs, there is something seriously wrong.

> And `Popen` is a context manager, so I think something like this should be possible:

Definitely no timeouts to kill the process should be used in the `multi` case (or we extend the definition of the run jobs to add it). I need this to run the hid-tools test suite, which can take 20 minutes by running sequentially.

Your solution would work for the `single` case, but not for the `multi` one where I need to get interactive results (think a test suite running or running `libinput debug-events` to check for `DEVICE_ADDED`). So I am not sure it adds a lot.

Reply:
> Definitely no timeouts to kill the process should be used in the multi case

ok, that bit can be fixed by handling the `TimeoutExpired` separate from the `KeyboardInterrupt`, it doesn't do anything other than time out.

Anyway, I think there's no point delaying the functionality itself, we can change this later if need be.